singularity-eos/eos/eos_builder.hpp

```diff
@@ -15,129 +15,130 @@
 #ifndef _SINGULARITY_EOS_EOS_EOS_BUILDER_HPP_
 #define _SINGULARITY_EOS_EOS_EOS_BUILDER_HPP_
 
-#include <map>
-#include <mpark/variant.hpp>
+#include <utility>
+
 #include <ports-of-call/portability.hpp>
-#include <singularity-eos/eos/eos.hpp>
-#include <string>
-#include <unordered_set>
 
-namespace singularity {
+#include <singularity-eos/base/variadic_utils.hpp>
+#include <singularity-eos/eos/eos_variant.hpp>
 
+namespace singularity {
 namespace EOSBuilder {
 
-// TODO(JMM): This could be strings? Or at least use a macro so we
-// only have to write these down once?
-// strings might allow us to create these structrs more easily
-// in the host code.
-enum class EOSType {
-  IdealGas,
-  Gruneisen,
-  JWL,
-  DavisReactants,
-  DavisProducts
-#ifdef SINGULARITY_USE_SPINER_WITH_HDF5
-  ,
-  SpinerEOSDependsRhoT,
-  SpinerEOSDependsRhoSie,
-  StellarCollapse
-#endif
-};
-enum class EOSModifier { Scaled, Shifted, Relativistic, UnitSystem, BilinearRamp };
-
-// evil type erasure
-using param_t = mpark::variant<bool, int, Real, std::string>;
-using params_t = std::map<std::string, param_t>;
-using modifiers_t = std::map<EOSModifier, params_t>;
-const params_t NO_PARAMS;
-
-// TODO: Extend if needed
-const std::unordered_set<EOSType> modifiable({EOSType::IdealGas
-#ifdef SINGULARITY_USE_SPINER_WITH_HDF5
-                                              ,
-                                              EOSType::SpinerEOSDependsRhoT,
-                                              EOSType::SpinerEOSDependsRhoSie,
-                                              EOSType::StellarCollapse
-#endif
-});
-bool isModifiable(EOSType t);
-EOS buildEOS(EOSType type, params_t base_params, modifiers_t modifiers);
-inline auto buildEOS(EOSType type, params_t base_params) {
-  modifiers_t modifiers;
-  return buildEOS(type, base_params, modifiers);
-}
+// Just import the variadic utils namespace int the EOSBuilder
+// namespace.
+using namespace variadic_utils;
 
-template <typename T>
-EOS makeUnitSystem(T &&eos, bool use_length_time, Real rho_unit, Real sie_unit,
-                   Real temp_unit, Real time_unit, Real mass_unit, Real length_unit) {
-  if (use_length_time) {
-    return UnitSystem<T>(std::move(eos), eos_units_init::length_time_units_init_tag,
-                         time_unit, mass_unit, length_unit, temp_unit);
-  } else {
-    return UnitSystem<T>(std::move(eos), eos_units_init::thermal_units_init_tag, rho_unit,
-                         sie_unit, temp_unit);
-  }
+template <template <class...> typename Mod, typename U, typename... Ts>
+using IsModifiable =
+    conjunction<contains<U, Ts...>::value, contains<Mod<U>, Ts...>::value>;
+template <template <class...> typename Mod, typename U, typename... Ts>
+constexpr bool is_modifiable(const Variant<Ts...> &var) {
```

Comment:
Could this function arg be replaced with a type list rather than a `Variant`?

Reply:
Sure---though I would suggest adding an overload with a type list version rather than getting rid of the variant version... because I was thinking maybe the variant version might be useful in userspace?

Reply:
Done. Added an overload that takes a type list. I recommend using that one for internals, but I think the variant version is useful for userspace.